app/embedder.py

```diff
@@ -56,62 +54,82 @@ def __init__(self, bert: str):
                 vocab_file=vocab_file,
                 do_lower_case=do_lower_case)
 
-    def GetEmbedding(self, seq: str) -> np.array:
+    def GetEmbedding(self, seqs: List[str]) -> Dict[str, np.array]:
+        result: Dict[str, np.array] = {}
+        undone_seqs: Set[str] = set()
+
         # fetch from cache
-        obj = Embedding.m.get(bert=self.bert, seq=seq)
-        if obj:
-            self.logger.info("Using embedding matrix fetched from MongoDB...")
-            # convert list back from list(floats) to np.array
-            matrix = np.array(obj.embedding)
-            self.logger.debug(matrix)
-            return matrix
-
-        self.logger.info("Embedding matrix for bert=%s, seq=%s "
-                         "not found in cache. Generating..." %
-                         (self.bert, seq))
-        matrix = self._buildEmbedding(seq)
-
-        # convert npArray to list for storage in MongoDB
-        l_matrix = matrix.tolist()
-        e = Embedding.make(dict(bert=self.bert, seq=seq, embedding=l_matrix))
-        e.m.save()
-        self.logger.info("Embedding matrix stored in MongoDB.")
-        return matrix
-
-    def _buildEmbedding(self, seq: str) -> np.array:
-        tokens = self.tokenizer.tokenize(seq)
-        if len(tokens) > MAX_SEQ_LENGTH - 2:
-            tokens = tokens[:(MAX_SEQ_LENGTH - 2)]
-
-        # CLS and SEP is a relic of multi-sentence pre-training
-        tokens = ["[CLS]"] + tokens + ["[SEP]"]
-
-        # segment_ids refers again to multi-sentence pre-training
-        # TODO: maybe scrap
-        segment_ids = [0] * len(tokens)
-
-        # per word, get IDs from vocab file
-        # e.g. [980234, 8792450, 132, 5002]
-        input_ids = self.tokenizer.convert_tokens_to_ids(tokens)
-
-        # The mask has 1 for real tokens and 0 for padding tokens. Only real
-        # tokens are attended to.
-        # e.g. [1, 1, 1, 1]
-        input_mask = [1] * len(input_ids)
-
-        # Zero-pad up to the sequence length.
-        padding = [0] * (MAX_SEQ_LENGTH - len(input_ids))  # [0, 0, 0, ...]
-        input_ids += padding  # [980234, 8792450, 132, 5002, 0, 0, 0, ...]
-        input_mask += padding  # [1, 1, 1, 1, 0, ...]
-        segment_ids += padding  # [0, 0, 0, ...]
-
-        assert len(input_ids) == MAX_SEQ_LENGTH
-        assert len(input_mask) == MAX_SEQ_LENGTH
-        assert len(segment_ids) == MAX_SEQ_LENGTH
-
-        bert_inputs = dict(input_ids=tf.expand_dims(input_ids, 0),
-                           input_mask=tf.expand_dims(input_mask, 0),
-                           segment_ids=tf.expand_dims(segment_ids, 0))
+        for i, seq in enumerate(seqs):
```

Comment:
I have a slight preference for having one outer loop of `if cached, return otherwise calculate` per sequence, as opposed to trying to fetch all from cache, then calculating all not-cached embeddings.

Reply:
TensorFlow is faster with batched operations, though.